### PR TITLE
V1.2.0

### PR DESCRIPTION
**Pull request by ottokruse**
*Issue #, if available:* N/A

*Description of changes:*
- nicer "human readable" error pages when L@E encounters an error
- nonce signing
- export all relevant outputs
- validate JSON config before updating function code
- added optional logging (useful during development)
- parseAuth detects when you are already signed in now (in case of parallel sign-ins, in e.g. separate browser tabs)
- refactored several pieces of code to make them more intent revealing as to help other possible contributors
- made L@E functions unit testable by pulling in getConfig into the handler

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rpattcorner** [comment]
Exciting.  I'm testing the changes to the little demo csp ... do you need me to test against this?

**Comment by ottokruse** [comment]
That would be great @rpattcorner !

**Comment by rpattcorner** [comment]
Can I modify the location to point to this?  Different ARN?  SemanticVersion?
```
      Location:
        ApplicationId: arn:aws:serverlessrepo:us-east-1:520945424137:applications/cloudfront-authorization-at-edge
        SemanticVersion: 1.1.0
``` being current

**Comment by rpattcorner** [comment]
```
Number of errors found: 1. Resource with id [LambdaEdgeProtection] is invalid. Application arn:aws:serverlessrepo:us-east-1:520945424137:applications/cloudfront-authorization-at-edge does not have a version 1.2.0. Rollback requested by user.
```
So I guess not.  I'm not set up for a full build, so maybe later.

**Comment by ottokruse** [comment]
Ah yes 1.2.0 is not on the SAR yet. Once this PR is merged I will publish it to the SAR

**Comment by rpattcorner** [comment]
Re `export all relevant outputs`, the logical next step for the `auth-only` template is to create a Cognito Identity Pool linked to the existing User Pool, and a related IAM Role that will give a generic SPA permissions to do whatever authorized operations it requires, e.g. call a Lambda, access a bucket, etc.  

I think the current outputs can support that scenario, which I'm seeing (1.1) as

- Cognito User Pool ID: currently available as UserPoolId
- Cognito App Client ID: currently available as ClientId

Hopefully that looks right to you, and these outputs will be maintained in the 1.2 release.  Is there anything else needed in outputs that would support the scenario we have in mind?

Also is there some reason we don't have an output for the cloudfront URL (dxxxxx.cloudfront.net)  to dynamically insert in the user pool Callback and Signout URLs?  For simple testing that seems much easier than a custom domain with the attendant certificate problems -- so I'm guessing there are alligators in that swamp?

**Comment by rpattcorner** [comment]
I think that we'll need the User Pool ARN, not just the ID.  For example in a stack run I see:

image

but the call to create a Cognito idP seems to want a whole arn, like:
```
   The provider name for an Amazon Cognito user pool. For example, cognito-idp.us-east-1.amazonaws.com/us-east-1_123456789. ```

**Comment by ottokruse** [comment]
> the logical next step for the auth-only template is to create a Cognito Identity Pool 

Well, that or the use of API Gateway, or AppSync, with JWT authorizers. This depends on developer trade offs, each options has its pros/cons.

> dynamically insert in the user pool Callback and Signout URL

The problem with this is, that there is a circular dependency in this scenario. CloudFront can only be created *after* the serverless app was initiated. The solution here is I think to carve out the Cognito user pool from the serverless app too, and create it beforehand. As soon as I have some bandwidth I will implement #19 to enable that.

**Comment by rpattcorner** [comment]
Great. Do you think 1.2 can return the full user pool arn any time soon?  Or do i need to work around?  Creating the pool outside the App would work too, but I’d think would take longer as you say

File: src/cfn-custom-resources/lambda-code-update/index.ts [hunk 1]
```diff
@@ -15,31 +15,32 @@ import { resolve } from 'path';
...
     const { data } = await axios.get(Code!.Location!, { responseType: 'arraybuffer' });
-
-    // Extract Lambda zip contents to temporary folder, add configuration.json, and rezip
     const lambdaZip = new Zip(data);
     console.log('Lambda zip contents:', lambdaZip.getEntries().map(entry => entry.name));
     console.log('Adding (fresh) configuration.json ...');
     const tempDir = mkdtempSync('/tmp/lambda-package');
     lambdaZip.extractAllTo(tempDir, true);
-    writeFileSync(resolve(tempDir, 'configuration.json'), Buffer.from(configuration));
+    writeFileSync(resolve(tempDir, 'configuration.json'), Buffer.from(JSON.stringify(config)));
```
Comment (karthikvadla) [inline src/cfn-custom-resources/lambda-code-update/index.ts:34]: `JSON.stringify(config`) -> `JSON.stringify(config, null, 2)`

File: src/lambda-edge/shared/shared.ts [hunk 2]
```diff
@@ -47,27 +49,118 @@ interface ConfigFromDisk {
...
 }
 
+let CONFIG: Config | undefined = undefined;
+
+export function setConfig(config?: Config) {
+    if (!process.env.IS_TEST_MODE) {
+        throw new Error('This method is intended to be used in test mode only: env var IS_TEST_MODE needs to be set');
+    }
+    CONFIG = config;
+}
```
Comment (karthikvadla) [inline src/lambda-edge/shared/shared.ts:123]: I'm not sure, if need this specific function for testing purpose.
With jest.mock we can mock what we needed.
```
    const config = {
        clientId: 'clientId',
        oauthScopes: ['oauthScope1'],
        cognitoAuthDomain: 'cognitoAuthDomain',
        redirectPathSignIn: 'redirectPathSignIn',
        redirectPathAuthRefresh: 'redirectPathAuthRefresh',
        cookieSettings: {nonce: 'test_nonce'},
        mode: 'mode',
        ...
    };
    fs.readFileSync.mockImplementation(() => {
        return config;
    });
```

So, getConfig() should still get us config.

File: src/lambda-edge/refresh-auth/index.ts [hunk 3]
```diff
@@ -3,69 +3,86 @@
...
 import { CloudFrontRequestHandler } from 'aws-lambda';
-import { getConfig, extractAndParseCookies, getCookieHeaders, httpPostWithRetry, createErrorHtml } from '../shared/shared';
-
-const { clientId, oauthScopes, cognitoAuthDomain, cookieSettings, mode, cloudFrontHeaders, clientSecret } = getConfig();
+import { getConfig, extractAndParseCookies, generateCookieHeaders, httpPostWithRetry, createErrorHtml } from '../shared/shared';
 
+let CONFIG: ReturnType<typeof getConfig>;
 
 export const handler: CloudFrontRequestHandler = async (event) => {
+    if (!CONFIG || process.env.IS_TEST_MODE) {
```
Comment (karthikvadla) [inline src/lambda-edge/refresh-auth/index.ts:11]: we can remove this `process.env.IS_TEST_MODE` if setConfig() is removed.

File: src/cfn-custom-resources/lambda-code-update/index.ts (hunk 1 above)
Comment (ottokruse) [reply src/cfn-custom-resources/lambda-code-update/index.ts:34]: Yeah good idea

File: src/lambda-edge/shared/shared.ts (hunk 2 above)
Comment (ottokruse) [reply src/lambda-edge/shared/shared.ts:123]: Allright, I'll remove that

File: src/lambda-edge/refresh-auth/index.ts (hunk 3 above)
Comment (ottokruse) [reply src/lambda-edge/refresh-auth/index.ts:11]: OK will do

**Comment by ottokruse** [comment]
> So it would be nice to return the ProviderName from the UserPool, but can be worked around with a synthetic per above

TBH I don't dislike the synthetic and think the IdentyPool declaration is the right place to contruct it - reveals its intent

**Pull request merged**